File: tests/test_yaml.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <cstdlib>

int main(int argc, char* argv[]) {
    // Try to find reference/meta.yaml - check multiple locations
    std::string meta_path;
    
    // Build search paths
    std::vector<std::string> search_dirs;
    
    // From executable directory
    if (argc > 0 && argv[0]) {
        std::string exe_path = argv[0];
        size_t pos = exe_path.find_last_of("/\\");
        if (pos != std::string::npos) {
            std::string exe_dir = exe_path.substr(0, pos);
            search_dirs.push_back(exe_dir + "/reference");
            search_dirs.push_back(exe_dir + "/../tests/reference");
            search_dirs.push_back(exe_dir + "/../../tests/reference");
        }
    }
    
    // Relative paths
    search_dirs.push_back("tests/reference");
    search_dirs.push_back("reference");
    
    for (const auto& dir : search_dirs) {
        std::string candidate = dir + "/meta.yaml";
        std::ifstream test_file(candidate);
        if (test_file.good()) {
            meta_path = candidate;
            break;
        }
    }
    
    if (meta_path.empty()) {
        std::cerr << "reference/meta.yaml not found" << std::endl;
        return 1;
    }
    
    std::ifstream file(meta_path);
    std::stringstream buffer;
    buffer << file.rdbuf();
    std::string payload = buffer.str();
    
    std::string raw_clash_proxies;
    std::string raw_clash_names;
    
    size_t p_start = payload.find("proxies:");
    if (p_start != std::string::npos) {
        p_start += 8; // skip 'proxies:'
        if (p_start < payload.length() && payload[p_start] == '\r') p_start++;
        if (p_start < payload.length() && payload[p_start] == '\n') p_start++;
        
        size_t next_section = std::string::npos;
        size_t search_pos = p_start;
        while ((search_pos = payload.find('\n', search_pos)) != std::string::npos) {
            search_pos++;
            if (search_pos < payload.length() && isalpha((unsigned char)payload[search_pos])) {
                next_section = search_pos;
                break;
            }
        }
        size_t p_end = (next_section != std::string::npos) ? next_section : payload.length();
        
        std::string block = payload.substr(p_start, p_end - p_start);
        
        std::string filtered_block;
        size_t search_start = 0;
        while (true) {
            size_t name_pos = block.find("- name:", search_start);
            if (name_pos == std::string::npos) break;
            
            size_t node_start = name_pos;
            while (node_start > 0 && (block[node_start - 1] == ' ' || block[node_start - 1] == '\t')) {
                node_start--;
            }
            
            size_t next_name_pos = block.find("- name:", name_pos + 7);
            size_t next_node_start = block.length();
            if (next_name_pos != std::string::npos) {
                next_node_start = next_name_pos;
                while (next_node_start > node_start && (block[next_node_start - 1] == ' ' || block[next_node_start - 1] == '\t')) {
                    next_node_start--;
                }
            }
            
            std::string node_str = block.substr(node_start, next_node_start - node_start);
            
            // Drop unsupported xhttp nodes
            if (node_str.find("xhttp") == std::string::npos) {
                filtered_block += node_str;
                
                size_t n_pos = node_str.find("- name:");
                size_t end_line = node_str.find('\n', n_pos);
                if (end_line == std::string::npos) end_line = node_str.length();
                size_t q_start = node_str.find_first_not_of(" \t'\"", n_pos + 7);
                size_t q_end = node_str.find_last_not_of(" \t\r\n'\"", end_line - 1);
                if (q_start != std::string::npos && q_end != std::string::npos && q_start <= q_end) {
                    raw_clash_names += "      - \"" + node_str.substr(q_start, q_end - q_start + 1) + "\"\n";
                }
            }
            search_start = next_name_pos;
        }
        raw_clash_proxies += filtered_block;
    }
    
    std::cout << "--- RAW CLASH PROXIES ---\n";
    std::cout << raw_clash_proxies << "\n";
    std::cout << "--- RAW CLASH NAMES ---\n";
    std::cout << raw_clash_names << "\n";
    return 0;
}

File: tests/test_yaml.py
import os
import sys

def test():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    candidates = [
        os.path.join(script_dir, 'reference', 'meta.yaml'),
        os.path.join(script_dir, '..', 'reference', 'meta.yaml'),
        'tests/reference/meta.yaml',
        'reference/meta.yaml'
    ]
    meta_path = None
    for c in candidates:
        if os.path.isfile(c):
            meta_path = c
            break

    if not meta_path:
        print("Error: meta.yaml not found", file=sys.stderr)
        sys.exit(1)

    with open(meta_path, 'r', encoding='utf-8') as f:
        payload = f.read()

    p_start = payload.find("proxies:")
    if p_start == -1: return
    p_start += 8
    if p_start < len(payload) and payload[p_start] == '\r': p_start += 1
    if p_start < len(payload) and payload[p_start] == '\n': p_start += 1

    next_section = -1
    search_pos = p_start
    while True:
        search_pos = payload.find('\n', search_pos)
        if search_pos == -1: break
        search_pos += 1
        if search_pos < len(payload) and payload[search_pos].isalpha():
            next_section = search_pos
            break

    p_end = next_section if next_section != -1 else len(payload)
    block = payload[p_start:p_end]

    filtered_block = ""
    raw_clash_names = ""
    search_start = 0

    while True:
        name_pos = block.find("- name:", search_start)
        if name_pos == -1: break

        node_start = name_pos
        while node_start > 0 and block[node_start - 1] in ' \t':
            node_start -= 1

        next_name_pos = block.find("- name:", name_pos + 7)
        next_node_start = len(block)
        if next_name_pos != -1:
            next_node_start = next_name_pos
            while next_node_start > node_start and block[next_node_start - 1] in ' \t':
                next_node_start -= 1

        node_str = block[node_start:next_node_start]

        if "xhttp" not in node_str:
            filtered_block += node_str
            n_pos = node_str.find("- name:")
            end_line = node_str.find('\n', n_pos)
            if end_line == -1: end_line = len(node_str)
            q_start = n_pos + 7
            while q_start < end_line and node_str[q_start] in " \t'\"": q_start += 1
            q_end = end_line - 1
            while q_end >= q_start and node_str[q_end] in " \t\r\n'\"": q_end -= 1
            if q_start <= q_end:
                raw_clash_names += "      - \"" + node_str[q_start:q_end + 1] + "\"\n"
        search_start = next_name_pos

    print("--- FILTERED BLOCK ---")
    print(filtered_block)
    print("--- NAMES ---")
    print(raw_clash_names)

test()
